### ISSUE TITLE
Dump and Load

### ISSUE THREAD

**Issue by BSVogler** [nest/nest-simulator#1755]
This issue is intended as a starting point for a discussion in the [hackathon](https://github.com/nest/nest-simulator/wiki/NEST-Hackathon-14.09-18.09-2020).

Following suggestion #1618, the problem of supporting iterative runs remains unsolved.
The current iterative workflow looks like this:

![suggestion for dump nest impl](https://user-images.githubusercontent.com/801818/93072372-9472ff80-f681-11ea-912b-de4492bee2d7.png)

The problem with the suggestion of a partial reset is that the point of a reset call is not defined.
The shared weights across episodes are one example of this workflow. There are other workflows where the starting conditions are changed instead of loading the weights.
### Suggestion
By specifying it with the introduction of an explicit dump command it could save a serialization in memory. This serialization could then later be loaded again.
The new workflow would then be reduced in computing time and requires less user code.
![suggestion for dump nest impl2](https://user-images.githubusercontent.com/801818/93073213-d94b6600-f682-11ea-899d-aa1542fff778.png)

### Specifics in dump and load
For this use case, the neurons and connections (connectome) are of importance as the network is fully reset each episode. The dumping could be incrementally extended to allow the dump to include different parts of the simulation.
The dump command could then look like this `nest.dump(["neuros,", "synapses", "RNG"])`.
`nest.load()` would only overwrite the loaded parameters, potentially throwing an error.
The parameters in the dump command make it explicit to the user which values are loaded.

### Further benefits
Calls to `ResetNetwork` prevent the usage of nest `prepare`+`run` resulting in a lot of overhead. If dumping and loading overwrites everything relevant to the computation the workflow could be further increased to 
Use `Prepare`, `Load`+`Run`, `Load`+`Run`…, `Cleanup`.

This requires that dumping is almost complete.
![suggestion for dump nest impl3](https://user-images.githubusercontent.com/801818/93073230-dea8b080-f682-11ea-818c-bb85d47686f4.png)

Depending on the serialization format it would also enable inspection or sharing of generated networks.

**Comment by jougs** [comment]
Many thanks to all for the (at least in my opinion) very nice discussion during the hackathon. @BSVogler can you update the issue as you go, i.e. add comments on the initial plan and progress and questions? Thanks!

**Comment by BSVogler** [comment]
Thank you for considering this suggestion. Below I want to sumamrize some discussion points.

It was generally agreed on that adding some form of serialization is a good idea, however, a proper implementation is very difficult to achieve. 

Because of performance reasons, a binary format saving the data structures was preferred.

The current idea is to add some PyNest interface with a basic, single-processor implementation as a starting point. This should be later then extended with lower-level code to increase the performance. It should also be extended to support distributed computing. I will be working towards a PR.

By using NestML different neuron models could be (de)serialized without explicit assignments from serialization to variables.

The reintroduction of a soft reset to support more iterative simulation runs was also considered as a possibility for solving this workflow requirement.

**Comment by jougs** [comment]
For completeness (and of course general amusement), please find attached two slide decks that describe earlier attempts of mine to solve the pertaining problem. Enjoy!

* [2004: Persistent Networks in NEST (using XDR)](https://github.com/nest/nest-simulator/files/5241455/Persistent.Networks.in.NEST.pdf)
* [2008: Persistent Networks in NEST (using HDF5)](https://github.com/nest/nest-simulator/files/5241458/PersistentNetworks.pdf)

**Comment by BSVogler** [comment]
I implemented a python implementation with tests ([Code in Gist](https://gist.github.com/BSVogler/fcd685f80bb8723fb6711b5b7453a35f)) and applied it to my reinforcement learning framework. It seems to work.

I encountered two problems which are currently solved with hacks in my library (no hacks in the intended nest library code):

- when using a custom model, the model needs to be defined before loading. This is not included in the dump.
For defining the stdp_dopamine_synapse the volume transmitter neuron id is needed. To obtain this id, you might want to load the dump first -> vicious circle
- neuron collections need to be updated, which were previously created each time. See #1821
